app/assets/javascripts/discourse/app/services/presence.js

```diff
@@ -543,11 +546,15 @@ export default class PresenceService extends Service {
           e.promiseProxy.resolve();
         }
       });
+
+      this._presenceDebounceMs = DEFAULT_PRESENCE_DEBOUNCE_MS;
     } catch (e) {
       // Put the failed events back in the queue for next time
       this._queuedEvents.unshift(...queue);
       if (e.jqXHR?.status === 429) {
-        // Rate limited. No need to raise, we'll try again later
+        // Rate limited
+        this._presenceDebounceMs =
+          e.jqXHR.responseJSON.extras.wait_seconds * 1000;
```

Comment:
If the 429 is caused by an intermediate load-balancer (on our hosting, HAProxy or NGINX), then this JSON may not be present. We should probably do something like

```suggestion
        const waitSeconds = e.jqXHR.responseJSON?.extras?.wait_seconds
        this._presenceDebounceMs = (waitSeconds || 10) * 1000;
```

Also... is there any situation where `jqXHR.responseJSON` would raise an exception? 🤔 (e.g. if the response content-type isn't JSON?)

Maybe we need a try/catch here as well?

Reply:
> is there any situation where jqXHR.responseJSON would raise an exception? 🤔 (e.g. if the response content-type isn't JSON?)

Had to check that, and though not really documented on the jQ site, `responseJSON` would just be `undefined` in case of any errors.